hotui_server/lib/hotui_server.dart

```diff
@@ -0,0 +1,256 @@
+  //
+  //    {"class":"_DemoItem","method":"build","library":"package:flutter_gallery/gallery/home.dart","methodBody":"Text('Hello, World')"}
+  Future<void> reloadMethod(String libraryId, String classId, String methodId,
+      String methodBody) async {
+    final Procedure procedure = await incrementalCompiler.compileExpression(
+        methodBody,
+        // This is not yet flexible for non-build expressions.
+        <String>['context'],
+        <String>['BuildContext'],
+        libraryId,
+        classId,
+        false);
+    component.transformChildren(
+        BodyReplacementTransformer(libraryId, classId, methodId, procedure));
+    final Library modifiedLibrary = component.libraries.firstWhere(
+        (Library library) => library.importUri.toString() == libraryId,
+        orElse: () => null);
+    if (modifiedLibrary == null) {
+      throw Exception('Could not find library with id: $libraryId');
+    }
+    final Component partialComponent = Component(libraries: <Library>[
+      modifiedLibrary,
+    ]);
+    partialComponent.unbindCanonicalNames();
+    partialComponent.computeCanonicalNames();
+    final IOSink sink =
+        File(path.join(outputDirectory.path, 'hotui.dill')).openWrite();
+    final BinaryPrinter printer =
+        LimitedBinaryPrinter(sink, (_) => true, false);
+    printer.writeComponentFile(partialComponent);
+    await sink.close();
+  }
+
+  // Expects newline denominated JSON object.
+  void onMessage(String line) {
+    Map<String, Object> message;
+    try {
+      message = json.decode(line);
+    } on FormatException catch (error) {
+      stderr.writeln(error.toString());
+      print(kFailedMessage);
+    }
+    final String libraryId = message['library'];
+    final String classId = message['class'];
+    final String methodId = message['method'];
+    final String methodBody = message['methodBody'];
+
+    reloadMethod(libraryId, classId, methodId, methodBody).then((void result) {
+      print(kReadyMessage);
+    }).catchError((dynamic error, StackTrace stackTrace) {
+      stderr.writeln(error.toString());
+      stderr.writeln(stackTrace.toString());
+      print(kFailedMessage);
+    });
+  }
+}
+
+/// Replaces the body of method [methodId] on class [classId] in library
+/// [libraryId] with [procedure].
+class BodyReplacementTransformer extends Transformer {
+  BodyReplacementTransformer(
+    this.libraryId,
+    this.classId,
+    this.methodId,
+    this.procedure,
+  );
+
+  final String libraryId;
+  final String classId;
+  final String methodId;
+  final Procedure procedure;
+
+  @override
+  Library visitLibrary(Library node) {
+    if (node.importUri.toString() == libraryId) {
+      return super.visitLibrary(node);
+    }
+    return node;
+  }
+
+  @override
+  Class visitClass(Class clazz) {
+    if (clazz.name == classId) {
+      return super.visitClass(clazz);
+    }
+    return clazz;
+  }
+
+  @override
+  TreeNode visitProcedure(Procedure node) {
```

Comment:
defensive programming nit: perhaps track whether you have matched the class, and library already when you hit the procedure and only perform the replacement when everything matches.
For example, maybe you can trigger visitProcedure in some context where the class has not matched yet (e.g. top level method or methods on a mixin).